python/xoscar/collective/backend/nccl_backend.py

```diff
@@ -0,0 +1,163 @@
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+                s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+                s.bind((host, port))
+                s.listen()
+                s.settimeout(0.5)
+                while self._run == 1:
+                    try:
+                        c_socket, addr = s.accept()
+                    except socket.timeout:
+                        continue
+
+                    t = threading.Thread(
+                        target=self._thread_request, args=(c_socket,), daemon=True
+                    )
+                    t.start()
+
+        def run(self, host=_store._DEFAULT_HOST, port=_store._DEFAULT_PORT):
+            # Run the TCP store in a different process
+            t = ExceptionAwareThreading(target=self._server_loop, args=(host, port))
+            t.start()
+            self._thread = t
+
+        def stop(self):
+            if _store._exit_mode:
+                return  # Prevent shutdown errors
+            if self._thread is not None:
+                # acquire the lock
+                self._lock.acquire()
+                self._run = 0
+                self._lock.release()
+                self._thread.join()
+
+    class XoscarNCCLBackend(cupyx.distributed.NCCLBackend):
+        """Interface that uses NVIDIA's NCCL to perform communications.
+
+        Args:
+            n_devices (int): Total number of devices that will be used in the
+                distributed execution.
+            rank (int): Unique id of the GPU that the communicator is associated to
+                its value needs to be `0 <= rank < n_devices`.
+            host (str, optional): host address for the process rendezvous on
+                initialization. Defaults to `"127.0.0.1"`.
+            port (int, optional): port used for the process rendezvous on
+                initialization. Defaults to `13333`.
+            use_mpi(bool, optional): switch between MPI and use the included TCP
+                server for initialization & synchronization. Defaults to `False`.
+        """
+
+        def __init__(
+            self,
+            n_devices,
+            rank,
+            tcpstore,
+            host=_store._DEFAULT_HOST,
+            port=_store._DEFAULT_PORT,
+            use_mpi=False,
+        ):
+            self._tcpstore = tcpstore
+            super().__init__(n_devices, rank, host, port, use_mpi)
+
+        def _init_with_tcp_store(self, n_devices, rank, host, port):
+            nccl_id = None
+            if rank == 0:
+                self._tcpstore.run(host, port)
+                nccl_id = nccl.get_unique_id()
+                # get_unique_id return negative values due to cython issues
+                # with bytes && c strings. We shift them by 128 to
+                # make them positive and send them as bytes to the proxy store
+                shifted_nccl_id = bytes([b + 128 for b in nccl_id])
```

Comment:
Is this just to_bytes? Can you use `numpy.ndarray.tobytes` or `struct.pack` to handle serialization?